hc_core/src/source_chain/memory/mod.rs

```diff
@@ -36,6 +36,25 @@ impl super::SourceChain for SourceChain {
     // @TODO - appending pairs should fail if hashes do not line up
     // @see https://github.com/holochain/holochain-rust/issues/31
     fn push(&mut self, pair: &super::Pair) {
+        let previous_pair = match pair.header.previous() {
+            Some(h) => self.get(h),
+            None => None,
+        };
+
+        // smoke test this pair in isolation, and check the hash reference against the top pair
+        if !(pair.validate() && self.pairs.first() == previous_pair.as_ref()) {
+            // we panic because no code path should attempt to append an invalid pair
+            panic!("attempted to push an invalid pair for this source chain");
+        }
+
+        // dry run an insertion against a clone and validate the outcome
+        let mut validation_chain = self.clone();
+        validation_chain.pairs.insert(0, pair.clone());
+        if !validation_chain.validate() {
+            // we panic because no code path should ever invalidate the chain
+            panic!("adding this pair would invalidate the source chain");
+        }
+
         // @TODO - inserting at the start of a vector is O(n), some other collection could be O(1)
         // @see https://github.com/holochain/holochain-rust/issues/35
         self.pairs.insert(0, pair.clone())
```

Comment:
I don't think the clone is necessary here. Or is it? Am I missing something?

Reply:
this was already in the code, not sure if it's relevant to this branch?

Reply:
Yeah... this can probably wait

Reply:
also i do get the compiler complaining at me about references when i drop it, so followup at #52